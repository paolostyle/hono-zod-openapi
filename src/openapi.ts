import { zValidator } from '@hono/zod-validator';
import type { Env, Hono, MiddlewareHandler } from 'hono';
import { every } from 'hono/combine';
import { createMiddleware } from 'hono/factory';
import { z } from 'zod';
import {
  createDocument,
  type ZodOpenApiObject,
  type ZodOpenApiOperationObject,
  type ZodOpenApiPathsObject,
  type ZodOpenApiRequestBodyObject,
  type ZodOpenApiResponseObject,
  type ZodOpenApiResponsesObject,
} from 'zod-openapi';
import type {
  AnyZ,
  EndpointDetails,
  Method,
  NormalizedRequestSchemas,
  PathsSchemas,
  RequestSchemas,
  ResponseParams,
  ResponseSchemas,
  StatusCodePrefix,
  StatusCodeWithWildcards,
  ValidationTarget,
  Values,
  ZodValidatorFn,
} from './types';

const OpenApiSymbol = Symbol();

function normalizeResponseSchemas<T extends AnyZ>(
  responseSchemas: T | ResponseParams<T> | Array<ResponseParams<T>>,
): Array<ResponseParams<T>> {
  if (Array.isArray(responseSchemas)) {
    return responseSchemas;
  }

  if ('schema' in responseSchemas) {
    return [responseSchemas];
  }

  return [
    {
      status: 200,
      schema: responseSchemas,
      description: '200 OK',
    },
  ];
}

export function createOpenApiMiddleware(
  zodValidator: ZodValidatorFn = zValidator,
  { validateResponse = false } = {},
) {
  return function openApi<
    ResS extends ResponseSchemas,
    ReqS extends RequestSchemas,
    E extends Env,
    P extends string,
  >(
    responseSchemas: ResS,
    requestSchemas: ReqS = {} as ReqS,
    endpointDetails: EndpointDetails = {},
  ): MiddlewareHandler<E, P, Values<ReqS>> {
    const normalizedResponseSchemas = normalizeResponseSchemas(responseSchemas);

    const validateResponseMiddleware = createMiddleware(async (c, next) => {
      await next();

      const response = c.res as Response;
      const responseParams = normalizedResponseSchemas.find((i) => {
        if (i.status === response.status) return true;
        const statusPrefix = `${response.status}`[0] as StatusCodePrefix;
        return i.status === statusPrefix + 'XX';
      });

      if (!responseParams) return;

      const { schema, validate, mediaType } = responseParams;
      const shouldValidate =
        validate || (validateResponse && validate !== false);
      const contentType = response.headers.get('content-type');

      if (!shouldValidate) return;

      if (mediaType && contentType !== mediaType) {
        throw new Error(
          'Content type does not match the response schema definition',
        );
      } else {
        responseParams.mediaType ??= contentType ?? 'application/json';
      }

      const copy = response.clone();
      const body = await response.json();

      schema.parse(body);

      c.res = copy;
    });

    const validators = Object.entries(requestSchemas)
      .map(([target, schemaOrParams]) => {
        const schema =
          schemaOrParams instanceof z.Schema
            ? schemaOrParams
            : schemaOrParams.validate !== false
              ? schemaOrParams.schema
              : null;

        if (!schema) return;

        return zodValidator(target as ValidationTarget, schema);
      })
      .filter((v) => !!v);

    if (validateResponse) {
      validators.push(validateResponseMiddleware);
    }

    const middleware = every(...validators);

    return Object.assign(middleware, {
      [OpenApiSymbol]: {
        request: requestSchemas,
        response: normalizedResponseSchemas,
        endpointDetails,
      },
    });
  };
}

export const openApi = createOpenApiMiddleware();

type Settings = {
  routeName?: string;
  addRoute?: boolean;
  overrides?: (
    paths: NonNullable<ZodOpenApiObject['paths']>,
  ) => Partial<ZodOpenApiObject>;
};

export function createOpenApiDocs(
  router: Hono,
  info: ZodOpenApiObject['info'],
  { addRoute = true, routeName = '/doc', overrides }: Settings = {},
  // fixme: return type
): Record<string, any> {
  const paths: ZodOpenApiPathsObject = {};

  router.routes
    .filter((route) => OpenApiSymbol in route.handler)
    .forEach((route) => {
      const { request, response, endpointDetails } = (route.handler as any)[
        OpenApiSymbol
      ] as PathsSchemas;

      const requestSchemas: NormalizedRequestSchemas = Object.fromEntries(
        Object.entries(request).map(
          ([key, value]) =>
            [key, value instanceof z.Schema ? value : value.schema] as const,
        ),
      );

      const responses = response.reduce<ZodOpenApiResponsesObject>(
        (
          acc,
          { description, schema, mediaType = 'application/json', status },
        ) => {
          acc[status as `${StatusCodeWithWildcards}`] = {
            description,
            content: {
              [mediaType]: {
                schema,
              },
            },
          } satisfies ZodOpenApiResponseObject;

          return acc;
        },
        {},
      );

      const requestBody: ZodOpenApiRequestBodyObject = {
        content: {
          'application/json': {
            schema: requestSchemas.json,
          },
        },
      };

      const operation: ZodOpenApiOperationObject = {
        requestParams: {
          cookie: requestSchemas.cookie,
          header: requestSchemas.header,
          query: requestSchemas.query,
          path: requestSchemas.param,
        },
        requestBody: requestSchemas.json && requestBody,
        responses,
        ...endpointDetails,
      };

      if (!(route.path in paths)) {
        paths[route.path] = {};
      }

      paths[route.path][route.method.toLowerCase() as Method] = operation;
    });

  const openApiDoc = createDocument({
    openapi: '3.1.0',
    info,
    paths,
    ...overrides?.(paths),
  });

  if (addRoute) {
    router.get(routeName, (c) => c.json(openApiDoc, 200));
  }

  return openApiDoc;
}
