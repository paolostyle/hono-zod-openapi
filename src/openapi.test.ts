import { Hono } from 'hono';
import { testClient } from 'hono/testing';
import { describe, expect, expectTypeOf, it, vi } from 'vitest';
import { z } from 'zod';
import { extendZodWithOpenApi } from 'zod-openapi';
import { createOpenApiDocs, createOpenApiMiddleware, openApi } from './openapi';

extendZodWithOpenApi(z);

const apiInfo = {
  title: 'Some API',
  version: '0.0.1',
};

describe('config: createOpenApiDocs', () => {
  it('correctly passes metadata to the document and creates the document under /doc route', async () => {
    const app = new Hono();

    createOpenApiDocs(app, apiInfo);

    const response = await app.request('/doc');
    const openApiSpec = await response.json();

    expect(response.status).toBe(200);
    expect(openApiSpec.openapi).toEqual('3.1.0');
    expect(openApiSpec.info).toEqual(apiInfo);
  });

  it('does not create route when addRoute = false', () => {
    const app = new Hono();

    createOpenApiDocs(app, apiInfo, { addRoute: false });

    expect(app.routes).toEqual([]);
  });

  it('creates a route under different name when routeName is provided', () => {
    const app = new Hono();

    createOpenApiDocs(app, apiInfo, { routeName: '/api-docs' });

    expect(app.routes[0].path).toEqual('/api-docs');
  });

  it('correctly adds additional attributes to the document through overrides', async () => {
    const app = new Hono();

    createOpenApiDocs(app, apiInfo, {
      overrides: () => ({
        servers: [{ url: 'https://api.example.com' }],
      }),
    });

    const response = await app.request('/doc');
    const openApiSpec = await response.json();

    expect(openApiSpec.servers).toEqual([{ url: 'https://api.example.com' }]);
  });
});

describe('paths: createOpenApiDocs', () => {
  it('simple', async () => {
    const app = new Hono().get(
      '/user',
      openApi(z.object({ hi: z.string() })),
      async (c) => {
        return c.json({ hi: 'string' }, 200);
      },
    );

    createOpenApiDocs(app, apiInfo);

    const response = await app.request('/doc');
    const openApiSpec = await response.json();

    expect(openApiSpec.paths).toMatchSnapshot();
  });

  it('object', async () => {
    const app = new Hono().get(
      '/user',
      openApi({
        schema: z.object({ hi: z.string() }).openapi({
          example: { hi: 'hello' },
        }),
        description: 'Great success',
        status: 200,
      }),
      async (c) => {
        return c.json({ hi: 'string' }, 200);
      },
    );

    createOpenApiDocs(app, apiInfo);

    const response = await app.request('/doc');
    const openApiSpec = await response.json();

    expect(openApiSpec.paths).toMatchSnapshot();
  });

  it('handles custom media type', async () => {
    const app = new Hono().get(
      '/user',
      openApi({
        schema: z.string(),
        description: 'Great success',
        mediaType: 'text/plain',
        status: 200,
      }),
      async (c) => {
        return c.text('yoooooooo', 200);
      },
    );

    createOpenApiDocs(app, apiInfo);

    const response = await app.request('/doc');
    const openApiSpec = await response.json();

    expect(
      openApiSpec.paths['/user'].get.responses['200'].content,
    ).toHaveProperty('text/plain');
  });

  it('properly normalizes response schema', async () => {
    const app1 = new Hono().get(
      '/example',
      openApi(z.object({ wow: z.string() })),
      async (c) => {
        return c.json({ wow: 'cool' }, 200);
      },
    );

    const doc1 = createOpenApiDocs(app1, apiInfo);

    const app2 = new Hono().get(
      '/example',
      openApi({
        status: 200,
        schema: z.object({ wow: z.string() }),
        description: '200 OK',
      }),
      async (c) => {
        return c.json({ wow: 'cool' }, 200);
      },
    );

    const doc2 = createOpenApiDocs(app2, apiInfo);

    const app3 = new Hono().get(
      '/example',
      openApi([
        {
          status: 200,
          schema: z.object({ wow: z.string() }),
          description: '200 OK',
        },
      ]),
      async (c) => {
        return c.json({ wow: 'cool' }, 200);
      },
    );

    const doc3 = createOpenApiDocs(app3, apiInfo);

    expect(doc1).toEqual(doc2);
    expect(doc2).toEqual(doc3);
  });

  it('properly handles multiple responses', async () => {
    const app = new Hono().get(
      '/example',
      openApi([
        {
          status: 200,
          schema: z.object({ wow: z.string() }),
          description: '200 OK',
        },
        {
          status: 400,
          schema: z.object({ error: z.string() }),
          description: '400 Bad Request',
        },
      ]),
      async (c) => {
        if (Math.random() > 0.5) {
          return c.json({ error: 'sike, wrong number' }, 200);
        }

        return c.json({ wow: 'cool' }, 200);
      },
    );

    createOpenApiDocs(app, apiInfo);

    const response = await app.request('/doc');
    const openApiSpec = await response.json();

    expect(openApiSpec.paths['/example'].get.responses).toMatchSnapshot();
  });
});

describe('default openApi middleware', () => {
  it('validates request body', async () => {
    const app = new Hono().post(
      '/user',
      openApi(z.object({ name: z.string() }), {
        json: z.object({ name: z.string() }),
      }),
      async (c) => {
        const body = c.req.valid('json');

        expectTypeOf(body).toEqualTypeOf<{ name: string }>();

        // @ts-expect-error c.req.valid() should only accept json
        const nonExistent = c.req.valid('cookie');
        // @ts-expect-error c.req.valid() should only accept json
        const nonExistent2 = c.req.valid('header');
        // @ts-expect-error c.req.valid() should only accept json
        const nonExistent3 = c.req.valid('query');
        // @ts-expect-error c.req.valid() should only accept json
        const nonExistent4 = c.req.valid('param');

        return c.json(body, 200);
      },
    );

    const client = testClient(app);
    const response = await client.user.$post({ json: { name: 'John' } });

    expect(response.status).toBe(200);
    expect(await response.json()).toEqual({ name: 'John' });

    // @ts-expect-error name should be a string
    const response2 = await client.user.$post({ json: { name: 123 } });
    expect(response2.status).toBe(400);
    expect(await response2.json()).toEqual({
      error: {
        issues: [
          {
            code: 'invalid_type',
            expected: 'string',
            message: 'Expected string, received number',
            path: ['name'],
            received: 'number',
          },
        ],
        name: 'ZodError',
      },
      success: false,
    });
  });

  it.each(['cookie', 'header', 'query', 'param', 'json'] as const)(
    'passes target %s and schema to validator',
    async (type) => {
      const validatorMock = vi.fn();
      const someSchema = z.object({ name: z.string() }).optional();
      const middleware = createOpenApiMiddleware(validatorMock);

      const app = new Hono().post(
        '/user',
        middleware(someSchema, {
          [type]: someSchema,
        }),
        async (c) => {
          return c.json({ name: 'John' }, 200);
        },
      );

      const client = testClient(app);
      await client.user.$post({} as any);

      expect(validatorMock).toHaveBeenCalledWith(type, someSchema);
    },
  );

  it('passes multiple targets and schemas to validator', async () => {
    const validatorMock = vi.fn();
    const someSchema = z.object({ name: z.string() });
    const middleware = createOpenApiMiddleware(validatorMock);

    const app = new Hono().post(
      '/user',
      middleware(z.object({ name: z.string() }), {
        cookie: someSchema,
        header: someSchema,
        query: someSchema,
        json: someSchema,
        param: someSchema,
      }),
      async (c) => {
        return c.json({ name: 'John' }, 200);
      },
    );

    const client = testClient(app);
    await client.user.$post({} as any);

    expect(validatorMock).toHaveBeenCalledWith('cookie', someSchema);
    expect(validatorMock).toHaveBeenCalledWith('header', someSchema);
    expect(validatorMock).toHaveBeenCalledWith('query', someSchema);
    expect(validatorMock).toHaveBeenCalledWith('json', someSchema);
    expect(validatorMock).toHaveBeenCalledWith('param', someSchema);
  });
});
