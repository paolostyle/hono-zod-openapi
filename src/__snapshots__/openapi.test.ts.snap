// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`paths: createOpenApiDocs > object 1`] = `
{
  "/user": {
    "get": {
      "responses": {
        "200": {
          "content": {
            "application/json": {
              "schema": {
                "example": {
                  "hi": "hello",
                },
                "properties": {
                  "hi": {
                    "type": "string",
                  },
                },
                "required": [
                  "hi",
                ],
                "type": "object",
              },
            },
          },
          "description": "Great success",
        },
      },
    },
  },
}
`;

exports[`paths: createOpenApiDocs > properly handles multiple responses 1`] = `
{
  "200": {
    "content": {
      "application/json": {
        "schema": {
          "properties": {
            "wow": {
              "type": "string",
            },
          },
          "required": [
            "wow",
          ],
          "type": "object",
        },
      },
    },
    "description": "200 OK",
  },
  "400": {
    "content": {
      "application/json": {
        "schema": {
          "properties": {
            "error": {
              "type": "string",
            },
          },
          "required": [
            "error",
          ],
          "type": "object",
        },
      },
    },
    "description": "400 Bad Request",
  },
}
`;

exports[`paths: createOpenApiDocs > simple 1`] = `
{
  "/user": {
    "get": {
      "responses": {
        "200": {
          "content": {
            "application/json": {
              "schema": {
                "properties": {
                  "hi": {
                    "type": "string",
                  },
                },
                "required": [
                  "hi",
                ],
                "type": "object",
              },
            },
          },
          "description": "200 OK",
        },
      },
    },
  },
}
`;
